_posts/2017-01-30-anim_compression_unity5.md

```diff
@@ -2,122 +2,118 @@
+
+Before we dig into what each compression setting does we must first briefly cover the data representations that [Unity 5](https://en.wikipedia.org/wiki/Unity_(game_engine)) uses internally.
+
+# Track Data Encoding
+The engine uses one of three encodings to represent an animation track regardless of the track data type (quaternion, vector, float, etc.):
+*   [Legacy Curve](http://nfrechette.github.io/2017/01/30/anim_compression_unity5/#legacy_curve)
+*   [Streaming Curve](http://nfrechette.github.io/2017/01/30/anim_compression_unity5/#streaming_curve)
+*   [Dense Curve](http://nfrechette.github.io/2017/01/30/anim_compression_unity5/#dense_curve)
+
+Rotation tracks are always encoded as four curves to represent a full quaternion (one curve per component). An obvious win here could be to instead encode rotations as quaternion logarithms or by dropping the quaternion **W** component or the largest component. This would of course immediately reduce the memory footprint for rotation tracks by **25%** at the expense of a few instructions to reconstruct the original quaternion.
+
+## Legacy Curve
+Legacy curves are a strange beast. The source data is sampled uniformly at a fixed interval such as **30 FPS** and is kept unsorted in full precision. Using discreet samples during decompression a [Hermite curve](https://en.wikipedia.org/wiki/Cubic_Hermite_spline) is constructed on the fly and interpolated. It is unclear to me how this format emerged but it has since been superseded by the other two and it is not frequently used.
+
+It must have been quite slow to decompress and should probably be avoided.
+
+## Streaming Curve
+Streaming curves are proper curves that use [Hermite coefficients](https://en.wikipedia.org/wiki/Cubic_Hermite_spline). A track is split into intervals and each interval is encoded as a distinct spline. This allows discontinuities between intervals. For example, a camera cut or teleporting the root in a cinematic. Each interval has a small header of **8** bytes and each control point is stored in full floating point precision plus an index. This is likely overkill. Full floating point precision is typically far too much for encoding rotations and using [simple quantization](http://nfrechette.github.io/2016/11/15/anim_compression_quantization/) to store them on 16 bits per component or less could provide significant memory savings.
+
+The resulting control points are sorted by time followed by track to render them in a cache as efficient as possible which is a very good thing. At decompression, a cursor or cache is used to avoid repeatedly searching for our control points when playback is continuous and predictable. For these two reasons streaming curves are very fast to decompress in the average use case.
```

Comment:
"render them in a cache as efficient" -> "render them as cache efficient" as in processor cache